TestFoundation/TestNSAttributedString.swift

```diff
@@ -251,12 +355,303 @@ class TestNSMutableAttributedString : XCTestCase {
+            ("test_replaceCharactersWithAttributedString", test_replaceCharactersWithAttributedString),
+            ("test_insert", test_insert),
+            ("test_append", test_append),
+            ("test_deleteCharacters", test_deleteCharacters),
+            ("test_setAttributedString", test_setAttributedString),
         ]
     }
     
     func test_initWithString() {
         let string = "Lorem 😀 ipsum dolor sit amet, consectetur adipiscing elit. ⌘ Phasellus consectetur et sem vitae consectetur. Nam venenatis lectus a laoreet blandit. ಠ_ರೃ"
         let mutableAttrString = NSMutableAttributedString(string: string)
         XCTAssertEqual(mutableAttrString.mutableString, NSMutableString(string: string))
-    }    
+    }
+    
+    func test_initWithStringAndAttributes() {
+        let string = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus consectetur et sem vitae consectetur. Nam venenatis lectus a laoreet blandit."
+        let attributes: [NSAttributedStringKey : Any] = [NSAttributedStringKey("attribute.placeholder.key") : "attribute.placeholder.value"]
+        
+        let mutableAttrString = NSMutableAttributedString(string: string, attributes: attributes)
+        XCTAssertEqual(mutableAttrString.mutableString, NSMutableString(string: string))
+        XCTAssertEqual(mutableAttrString.length, string.utf16.count)
+        
+        var range = NSRange()
+        let attrs = mutableAttrString.attributes(at: 0, effectiveRange: &range)
+        guard let value = attrs[NSAttributedStringKey("attribute.placeholder.key")] as? String else {
+            XCTAssert(false, "attribute value not found")
+            return
+        }
+        XCTAssertEqual(range.location, 0)
+        XCTAssertEqual(range.length, mutableAttrString.length)
+        XCTAssertEqual(value, "attribute.placeholder.value")
+        
+        let invalidAttribute = mutableAttrString.attribute(NSAttributedStringKey("invalid"), at: 0, effectiveRange: &range)
+        XCTAssertNil(invalidAttribute)
+        XCTAssertEqual(range.location, 0)
+        XCTAssertEqual(range.length, string.utf16.count)
+        
+        let attribute = mutableAttrString.attribute(NSAttributedStringKey("attribute.placeholder.key"), at: 0, effectiveRange: &range)
+        XCTAssertEqual(range.location, 0)
+        XCTAssertEqual(range.length, mutableAttrString.length)
+        guard let validAttribute = attribute as? NSString else {
+            XCTAssert(false, "attribute not found")
+            return
+        }
+        XCTAssertEqual(validAttribute, "attribute.placeholder.value")
+    }
+    
+    func test_initWithAttributedString() {
+        let string1 = "Lorem ipsum dolor sit amet, consectetur adipiscing elit."
+        let attrKey1 = NSAttributedStringKey("attribute.placeholder.key1")
+        let attrValue1 = "attribute.placeholder.value1"
+        let attrRange1 = NSRange(location: 0, length: string1.utf8.count)
```

Comment:
You could also use the new `NSRange` and `Range` conversions, e.g. https://github.com/apple/swift/blob/master/stdlib/public/SDK/Foundation/NSRange.swift#L180 and https://github.com/apple/swift/blob/master/stdlib/public/SDK/Foundation/NSRange.swift#L148

Reply:
Thanks for the review.

These `string.utf8.count` were mistakes, I meant `utf16` to get the `length` as `NSString` does, sorry for the confusion.

I did some cleanups:
- Use `NSMutableAttributedString.init(string:attributes:)` when possible, this reduces some uses of the string ranges
- Convert `NSRange` from `Range` when trying to get the full range of the string